core/state/statedb.go

```diff
@@ -1640,6 +1640,40 @@ func (s *StateDB) SlotInAccessList(addr common.Address, slot common.Hash) (addre
 	return s.accessList.Contains(addr, slot)
 }
 
+func (s *StateDB) ValidateKnownAccounts(knownAccounts types.KnownAccounts) error {
+	if knownAccounts == nil {
+		return nil
+	}
+
+	for k, v := range knownAccounts {
+		// check if the value is hex string or an object
+		switch {
+		case v.IsSingle():
+			trie, _ := s.StorageTrie(k)
+			if trie != nil {
+				actualRootHash := trie.Hash()
+				if *v.Single != actualRootHash {
+					return fmt.Errorf("invalid root hash for: %v root hash: %v actual root hash: %v", k, v.Single, actualRootHash)
+				}
+			} else {
+				return fmt.Errorf("Storage Trie is nil for: %v", k)
+			}
+		case v.IsStorage():
+			for slot, value := range v.Storage {
+				actualValue := s.GetState(k, slot)
+				if value != actualValue {
+					return fmt.Errorf("invalid slot value at address: %v slot: %v value: %v actual value: %v", k, slot, value, actualValue)
+				}
+			}
+		default:
+			trie, _ := s.StorageTrie(k)
+			return fmt.Errorf("invalid root hash for: %v root hash: %v actual root hash: %v", k, v.Single, trie.Hash())
```

Comment:
Curious to know how do we get `v.Single` if the case `v.IsSingle()` is false?

Reply:
Thanks for pointing out. Updated [here](https://github.com/maticnetwork/bor/pull/945/commits/0dd4a4eb3f9f78d25afb746bbff149ed556c4ae9).